### PR TITLE
Add Transaction syntax improvements

### PR DESCRIPTION
**Pull request by anupama-pathirage**
- Failed block is renamed into onretry
- Two handlers added for transaction - oncommit, onabort.

Sample syntax is as follows.
```ballerina
import ballerina.io;

function main (string[] args) {

    transaction with retries(4), oncommit(foo), onabort(bar){
        //...
        //sql update1
        //sql update2
        //...
    } onretry {
        io:println("Retry the Tx");
    }
}


function foo() {
    io:println("Tx Committed");
}


function bar() {
    io:println("Tx aborted");
}
```

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/BLangProgramRunner.java [hunk 1]
```diff
@@ -88,6 +94,15 @@ public static void runMain(ProgramFile programFile, String[] args) {
         }
     }
 
+    private static void initializeDistributeTransactionStatus(ProgramFile programFile) {
+        boolean distributedTxEnabled = false; //TODO:Default will be true. Read from new VMOptions
+        String distributedTxEnabledProp = System.getProperty(Constants.DISTRIBUTED_TRANSACTION_ENABLED);
+        if (distributedTxEnabledProp != null && !distributedTxEnabledProp.isEmpty()) {
```
Comment (lafernando) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/BLangProgramRunner.java]: If you're checking to see if it's not empty also .. that means .. we expect a value from it .. such as "true" or "false"? .. if that's the case, we have to boolean parse that value and set it. In this way, any value put into the system property, the flag will be true.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 2]
```diff
@@ -2630,7 +2634,11 @@ private static void createNewStreamlet(WorkerExecutionContext ctx, int[] operand
         sf.refRegs[i] = streamlet;
     }
 
-    private static void beginTransaction(WorkerExecutionContext ctx, int transactionBlockId, int retryCountRegIndex) {
+    private static void beginTransaction(WorkerExecutionContext ctx, int transactionBlockId, int retryCountRegIndex,
+            int committedFuncIndex, int abortedFuncIndex) {
+        //If global tx enabled, it is managed via transaction
```
Comment (lafernando) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Put a more clearer comment  here ..

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 3]
```diff
@@ -2642,21 +2650,46 @@ private static void beginTransaction(WorkerExecutionContext ctx, int transaction
...
         LocalTransactionInfo localTransactionInfo = ctx.getLocalTransactionInfo();
         boolean isInitiator = true;
         if (localTransactionInfo == null) {
-            BValue[] returns = notifyTransactionBegin(ctx, null,  null, transactionBlockId,
-                            TransactionConstants.DEFAULT_COORDINATION_TYPE, isInitiator);
-            BStruct txDataStruct = (BStruct) returns[0];
-            String globalTransactionId = txDataStruct.getStringField(1);
-            String protocol = txDataStruct.getStringField(2);
-            String url = txDataStruct.getStringField(3);
+            String globalTransactionId = "";
```
Comment (lafernando) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: We shouldn't initialize a variable with a value when its value is surely assigned / should be assigned in the following code.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 4]
```diff
@@ -2642,21 +2650,46 @@ private static void beginTransaction(WorkerExecutionContext ctx, int transaction
...
+            String protocol = "";
```
Comment (lafernando) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Why are these variables initialized with empty strings? .. shouldn't they be null if they are not assigned a value.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/statements/TransactionNode.java [hunk 5]
```diff
@@ -25,14 +25,22 @@
 public interface TransactionNode extends StatementNode {
```
Comment (lafernando) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/statements/TransactionNode.java]: Do put an interface level comment.

Comment (anupama-pathirage) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/statements/TransactionNode.java]: Fixed with 40fa23b1031395b8a645dd6c6f16ee5db672b2e4

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java [hunk 6]
```diff
@@ -59,7 +59,7 @@
         this.pkgBuilder = new BLangPackageBuilder(context, compUnit);
         this.diagnosticSrc = diagnosticSource;
         String distributedTxEnabledProp = System.getProperty(DISTRIBUTED_TRANSACTION_ENABLED);
-        if (distributedTxEnabledProp != null && !distributedTxEnabledProp.isEmpty()) {
+        if (distributedTxEnabledProp != null) {
```
Comment (lafernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: This and the above approach (BLangProgramRunner) of checking is dtx enabled is different. Bring the implementation to a common function and use it from all the places.

Comment (anupama-pathirage) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: Fixed with 052fbe446ddc7a23a4a7f62411a3875bfe64d96f

**Pull request merged**